memiavl/tree.go

```diff
@@ -16,54 +18,70 @@ type Tree struct {
 	version uint32
 	// root node of empty tree is represented as `nil`
 	root Node
-
+	// write ahead log to store changesets for each block
+	bwal           blockWAL
 	initialVersion uint32
 }
 
 // NewEmptyTree creates an empty tree at an arbitrary version.
-func NewEmptyTree(version int64) *Tree {
+func NewEmptyTree(version uint64, pathToWAL string) *Tree {
 	if version >= math.MaxUint32 {
 		panic("version overflows uint32")
 	}
-	return &Tree{version: uint32(version)}
+
+	wal, err := newBlockWAL(DefaultPathToWAL, version, nil)
+	if err != nil {
+		panic(err)
+	}
+	return &Tree{version: uint32(version), bwal: wal}
 }
 
 // New creates an empty tree at genesis version
-func New() *Tree {
-	return NewEmptyTree(0)
+func New(pathToWAL string) *Tree {
+	return NewEmptyTree(0, pathToWAL)
 }
 
 // New creates a empty tree with initial-version,
 // it happens when a new store created at the middle of the chain.
-func NewWithInitialVersion(initialVersion int64) *Tree {
+func NewWithInitialVersion(initialVersion int64, pathToWAL string) *Tree {
 	if initialVersion >= math.MaxUint32 {
 		panic("version overflows uint32")
 	}
-	tree := New()
+	tree := New(pathToWAL)
 	tree.initialVersion = uint32(initialVersion)
 	return tree
 }
 
 // NewFromSnapshot mmap the blob files and create the root node.
-func NewFromSnapshot(snapshot *Snapshot) *Tree {
+func NewFromSnapshot(snapshot *Snapshot, pathToWAL string) *Tree {
 	if snapshot.IsEmpty() {
-		return NewEmptyTree(int64(snapshot.Version()))
+		return NewEmptyTree(uint64(snapshot.Version()), pathToWAL)
+	}
+
+	wal, err := newBlockWAL(pathToWAL, uint64(snapshot.Version()), nil)
+	if err != nil {
+		panic(err)
 	}
+
 	return &Tree{
 		version: snapshot.Version(),
 		root:    snapshot.RootNode(),
+		bwal:    wal,
 	}
 }
 
 // ApplyChangeSet apply the change set of a whole version, and update hashes.
+// Returns hash, new version, and potential error.
 func (t *Tree) ApplyChangeSet(changeSet *iavl.ChangeSet, updateHash bool) ([]byte, int64, error) {
+	t.bwal.BlockChangeset = *changeSet
 	for _, pair := range changeSet.Pairs {
 		if pair.Delete {
 			t.remove(pair.Key)
 		} else {
 			t.set(pair.Key, pair.Value)
 		}
 	}
+	t.bwal.Flush()
```

Comment:
Should we just call `t.bwal.writeBlockChanges ` directly here?

Reply:
`Flush()` does that internally. but I see what you mean. I think I can just get rid of `BlockChangeset` in wal struct, because we do not need it anymore, since `set/remove` are not exposed

Reply:
yeah, that'd be simpler.

Reply:
ok, changed that, now changesets can be directly provided as `Flush` args

Reply:
I will also move `Flush` to be ran before all sets and removes